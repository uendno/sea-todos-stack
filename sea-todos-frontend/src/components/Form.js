import React, {Component} from 'react';
import PropTypes from 'prop-types';

export default class Form extends Component {
  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
  }

  state = {
    text: ''
  };

  onTextChange = (e) => {
    this.setState({
      text: e.target.value
    });
  };

  onKeyPress = (e) => {
    const {onSubmit} = this.props;
    const {text} = this.state;

    if (e.key === 'Enter' && text !== '') {
      onSubmit(text);
      this.setState({
        text: ''
      })
    }
  };

  render() {
    const {text} = this.state;

    return (
      <div>
        <input
          className="new-todo"
          value={text}
          placeholder="What needs to be done?"
          onChange={this.onTextChange}
          onKeyPress={this.onKeyPress}
        />
      </div>
    )
  }
}
