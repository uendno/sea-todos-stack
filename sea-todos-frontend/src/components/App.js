import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { showSuccessMessage } from 'utils/toast';
import Task from './Task';
import Form from './Form';
import Filter from './Filter';
import { getTasks, toggleTask, addTask } from '../actions/tasks';
import { getFilteredTaskListSelector } from '../reducers';

export class App extends Component {
  static propTypes = {
    getTasks: PropTypes.func.isRequired,
    toggleTask: PropTypes.func.isRequired,
    addTask: PropTypes.func.isRequired,
    tasks: PropTypes.array.isRequired,
  }

  async componentDidMount() {
    const {getTasks} = this.props;
    getTasks();
  }

  addTask = async (text) => {
    const {addTask} = this.props;
    await addTask(text);
    showSuccessMessage('Success!', 'Add new task successfully!');
  };

  render() {
    const {tasks, toggleTask} = this.props;
    return (
      <section className="todoapp">
        <header className="header">
          <h1>todos 2</h1>
          <Form onSubmit={this.addTask}/>
        </header>
        <section className="main">
          <ul className="todo-list">
            {tasks
              .map(task => (
              <Task
                key={task.id}
                task={task}
                toggleTask={toggleTask}
              />
            ))}
          </ul>
        </section>
        <footer className="footer">
          <Filter />
        </footer>
      </section>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    tasks: getFilteredTaskListSelector(state)
  }
};

const mapDispatchToProps = {
  getTasks,
  toggleTask,
  addTask
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
