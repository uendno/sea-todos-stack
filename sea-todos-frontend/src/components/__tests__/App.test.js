import React from 'react';
import {shallow} from 'enzyme';
import {App} from '../App';
import Form from '../Form';

describe('components/App', () => {
  let wrapper;
  let props;

  const setup = () => {
    wrapper = shallow(<App {...props} />)
  }

  beforeEach(() => {
    props = {
      getTasks: jest.fn(),
      toggleTask: jest.fn(),
      addTask: jest.fn(),
      tasks: [
        {
          id: 1
        },
        {
          id: 2
        },
        {
          id: 3
        }
      ]
    }
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should add a new task correctly', () => {
    setup();
    wrapper.find(Form).props().onSubmit();
    expect(props.addTask).toHaveBeenCalled();
  })
});
