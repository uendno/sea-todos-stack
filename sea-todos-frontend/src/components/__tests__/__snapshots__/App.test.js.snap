// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/App should render correctly 1`] = `
<section
  className="todoapp"
>
  <header
    className="header"
  >
    <h1>
      todos
    </h1>
    <Form
      onSubmit={[Function]}
    />
  </header>
  <section
    className="main"
  >
    <ul
      className="todo-list"
    >
      <Task
        key="1"
        task={
          Object {
            "id": 1,
          }
        }
        toggleTask={[MockFunction]}
      />
      <Task
        key="2"
        task={
          Object {
            "id": 2,
          }
        }
        toggleTask={[MockFunction]}
      />
      <Task
        key="3"
        task={
          Object {
            "id": 3,
          }
        }
        toggleTask={[MockFunction]}
      />
    </ul>
  </section>
  <footer
    className="footer"
  >
    <Connect(Filter) />
  </footer>
</section>
`;
